janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/service/GroupingAsyncQueryExecutionService.java

```diff
@@ -0,0 +1,433 @@
+import static org.janusgraph.diskstorage.cql.CQLConfigOptions.SLICE_GROUPING_LIMIT;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.COLUMN_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.COLUMN_COLUMN_NAME;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.KEY_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.KEY_COLUMN_NAME;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.LIMIT_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.SLICE_END_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.SLICE_START_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.VALUE_COLUMN_NAME;
+
+public class GroupingAsyncQueryExecutionService implements AsyncQueryExecutionService {
+
+    private final ResultFiller<Map<StaticBuffer, CompletableFuture<EntryList>>, SliceQuery, KeysGroup> SINGLE_QUERY_WITH_KEYS_GROUPING_FILLER;
+    private final ResultFiller<Map<StaticBuffer, CompletableFuture<EntryList>>, SliceQuery, List<StaticBuffer>> SINGLE_QUERY_WITHOUT_KEYS_GROUPING_FILLER;
+    private final ResultFiller<Map<SliceQuery, Map<StaticBuffer, CompletableFuture<EntryList>>>, QueryGroups, KeysGroup> MULTI_QUERY_WITH_KEYS_GROUPING_FILLER;
+    private final ResultFiller<Map<SliceQuery, Map<StaticBuffer, CompletableFuture<EntryList>>>, QueryGroups, List<StaticBuffer>> MULTI_QUERY_WITHOUT_KEYS_GROUPING_FILLER;
+
+    private final AsyncCQLSingleKeySliceFunction cqlSingleKeySliceFunction;
+    private final AsyncCQLSingleKeyMultiColumnFunction cqlSingleKeyMultiColumnFunction;
+    private final AsyncCQLMultiKeySliceFunction cqlMultiKeySliceFunction;
+    private final AsyncCQLMultiKeyMultiColumnFunction cqlMultiKeyMultiColumnFunction;
+    private final boolean sliceGroupingAllowed;
+    private final int sliceGroupingLimit;
+    private final boolean keysGroupingAllowed;
+    private final int keysGroupingLimit;
+    private final int keysGroupingMin;
+    private final GroupedExecutionStrategy groupedExecutionStrategy;
+
+    public GroupingAsyncQueryExecutionService(Configuration configuration,
+                                              final CQLStoreManager storeManager,
+                                              String tableName,
+                                              Function<Select, Select> addTTLFunction,
+                                              Function<Select, Select> addTimestampFunction,
+                                              CQLColValGetter singleKeyGetter,
+                                              CQLColValGetter multiKeysGetter) {
+
+        sliceGroupingLimit = Math.max(1, configuration.get(SLICE_GROUPING_LIMIT));
+        keysGroupingLimit = Math.max(1, configuration.get(KEYS_GROUPING_LIMIT));
+        keysGroupingMin = Math.max(2, configuration.get(KEYS_GROUPING_MIN));
```

Comment:
(nit) shall we throw exceptions or add a warning log when the user-specified number is less than 1 or 2?

Reply:
Make sense. I added warning in case user selects a value which is less than needed.